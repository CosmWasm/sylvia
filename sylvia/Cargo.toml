[package]
name = "sylvia"
version = { workspace = true }
authors = ["Bart≈Çomiej Kuras <bart.k@confio.gmbh>"]
edition = "2021"
description = "Framework for generating trait based CosmWasm smart contracts"
license = "Apache-2.0"
repository = "https://github.com/CosmWasm/sylvia"
homepage = "https://cosmwasm.com"
keywords = ["cosmwasm", "blockchain", "confio"]
categories = ["api-bindings", "wasm"]
readme = "../README.md"

[features]
mt = ["sylvia-derive/mt", "dep:cw-multi-test", "dep:anyhow"]
stargate = ["cosmwasm-std/stargate", "cw-multi-test/stargate"]
iterator = ["cosmwasm-std/iterator"]
cosmwasm_1_1 = ["cosmwasm-std/cosmwasm_1_1", "cw-multi-test/cosmwasm_1_1"]
cosmwasm_1_2 = [
    "cosmwasm-std/cosmwasm_1_2",
    "cw-multi-test/cosmwasm_1_2",
    "sylvia-derive/cosmwasm_1_2",
    "cosmwasm_1_1",
]
cosmwasm_1_3 = [
    "cosmwasm-std/cosmwasm_1_3",
    "cw-multi-test/cosmwasm_1_3",
    "cosmwasm_1_2",
]
cosmwasm_1_4 = [
    "cosmwasm-std/cosmwasm_1_4",
    "cw-multi-test/cosmwasm_1_4",
    "cosmwasm_1_3",
]
cosmwasm_2_0 = [
    "cosmwasm-std/cosmwasm_2_0",
    "cw-multi-test/cosmwasm_2_0",
    "cosmwasm_1_4",
]
cosmwasm_2_1 = [
    "cosmwasm-std/cosmwasm_2_1",
    "cw-multi-test/cosmwasm_2_1",
    "cosmwasm_2_0",
]
cosmwasm_2_2 = ["cosmwasm-std/cosmwasm_2_2", "cosmwasm_2_1"]

[dependencies]
sylvia-derive = { workspace = true }
cosmwasm-std = { workspace = true, features = ["staking"] }
cosmwasm-schema = { workspace = true }
schemars = { workspace = true }
serde = { workspace = true }
serde-cw-value = "0.7.0"
serde-json-wasm = "1.0.1"
konst = "0.3.15"
cw-multi-test = { workspace = true, features = ["staking"], optional = true }
anyhow = { workspace = true, optional = true }
cw-utils = { workspace = true }

[dev-dependencies]
cw-multi-test = { workspace = true, features = ["staking"] }
anyhow = { workspace = true }
cw-storage-plus = { workspace = true }
cw-utils = { workspace = true }
thiserror = { workspace = true }
trybuild = "1.0.101"
itertools = "0.13.0"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--generate-link-to-definition"]
