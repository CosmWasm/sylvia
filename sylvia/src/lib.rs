#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer.
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here.

pub mod builder;
pub mod ctx;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
pub use cw_utils;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value,
};
