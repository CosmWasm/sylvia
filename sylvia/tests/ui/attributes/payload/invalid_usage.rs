#![allow(unused_imports, unused_variables)]
use sylvia::ctx::{InstantiateCtx, ReplyCtx};
use sylvia::cw_std::{Addr, Binary, Response, StdResult};

pub mod used_on_context {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, reply_on=success)]
        fn reply(&self, #[sv::payload(raw)] _ctx: ReplyCtx) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

pub mod used_on_self {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, reply_on=success)]
        fn reply(
            #[sv::payload(raw)] &self,
            _ctx: ReplyCtx,
            payload: Binary,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

fn main() {}
