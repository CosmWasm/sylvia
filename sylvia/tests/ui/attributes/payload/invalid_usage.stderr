error: Invalid usage of Sylvia attribute.

         = note: First and second arguments of the method should be `self` and `ctx` respectively.
         = note: Unexpected attribute on `ctx` parameter.

  --> tests/ui/attributes/payload/invalid_usage.rs:22:25
   |
22 |         fn reply(&self, #[sv::payload(raw)] _ctx: ReplyCtx) -> StdResult<Response> {
   |                         ^

error: Missing payload parameter.

         = note: Expected at least one payload parameter at the end of parameter list.

  --> tests/ui/attributes/payload/invalid_usage.rs:22:12
   |
22 |         fn reply(&self, #[sv::payload(raw)] _ctx: ReplyCtx) -> StdResult<Response> {
   |            ^^^^^

error: Invalid usage of Sylvia attribute.

         = note: First and second arguments of the method should be `self` and `ctx` respectively.
         = note: Unexpected attribute on `self` parameter.

  --> tests/ui/attributes/payload/invalid_usage.rs:46:13
   |
46 |             #[sv::payload(raw)] &self,
   |             ^
